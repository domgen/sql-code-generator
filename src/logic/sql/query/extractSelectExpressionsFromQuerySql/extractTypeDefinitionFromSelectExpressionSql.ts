import { TypeDefinitionOfQuerySelectExpression } from '../../../../model';

export const extractTypeDefinitionFromSelectExpressionSql = ({ sql }: { sql: string }) => {
  // grab the source path
  const [_, sourcePath] = new RegExp(/^(\w+\.?\w*)(?:\s+(?:[\w\s]*))?$/).exec(sql) ?? []; // tslint:disable-line no-unused
  if (!sourcePath) throw new Error('could not extract source path from select expression sql; unexpected'); // fail fast

  // grab the alias, if any
  const [__, alias] = new RegExp(/(?:[\w\.]+)(?:\s+)(?:as)(?:\s+)(\w+)/g).exec(sql) ?? []; // tslint:disable-line no-unused

  // define the name, considering whether alias was given
  const name = alias ? alias : sourcePath.split('.').slice(-1)[0];

  // return the full definition
  return new TypeDefinitionOfQuerySelectExpression({ name, sourcePath });
};
