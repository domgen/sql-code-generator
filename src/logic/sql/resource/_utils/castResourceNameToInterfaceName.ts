import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
