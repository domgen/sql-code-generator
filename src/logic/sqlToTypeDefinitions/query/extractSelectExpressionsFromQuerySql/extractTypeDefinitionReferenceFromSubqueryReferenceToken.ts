import { TypeDefinitionReference } from '../../../../domain';
import { SqlSubqueryReference } from '../../../../domain/objects/SqlSubqueryReference';
import { getTokenForSqlSubqueryReference } from '../common/flattenSqlByReferencingAndTokenizingSubqueries/getTokenForSubqueryReference';
import { extractSelectExpressionsFromQuerySql } from './extractSelectExpressionsFromQuerySql';

export const extractTypeDefinitionReferenceFromSubqueryReferenceToken = ({
  subqueryReferenceToken,
  subqueries,
}: {
  subqueryReferenceToken: string;
  subqueries: SqlSubqueryReference[];
}): TypeDefinitionReference => {
  // 1. find the subquery reference object
  const subquery = subqueries.find(
    (subquery) =>
      getTokenForSqlSubqueryReference({ reference: subquery }) ===
      subqueryReferenceToken,
  );
  if (!subquery) {
    throw new Error(
      'subquery reference token found in select expression - but subquery reference definition not present. unexpected',
    ); // fail fast
  }

  // 2. get select expressions from subquery sql
  const cleanedSubquerySql = subquery.sql.replace(/^\(/, '').replace(/\)$/, ''); // if it has leading or closing paren, strip em
  const selectExpressions = extractSelectExpressionsFromQuerySql({
    sql: cleanedSubquerySql,
    inASubquery: true,
  }); // NOTE: this makes this recursive

  // 3. check that the subquery returns exactly one select expression, otherwise this is a syntax error and not valid sql
  if (selectExpressions.length < 1) {
    throw new Error(
      'subquery in select expression must return atleast one value.',
    );
  }
  if (selectExpressions.length > 1) {
    throw new Error(
      'subquery in select expression may not return more than one value. this is invalid sql',
    );
  }

  // 4. the reference type of this expression is the single reference type of that subquery
  return selectExpressions[0].typeReference;
};
