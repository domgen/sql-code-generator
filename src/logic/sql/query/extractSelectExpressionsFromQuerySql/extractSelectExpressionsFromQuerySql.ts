import { extractTypeDefinitionFromSelectExpressionSql } from './extractTypeDefinitionFromSelectExpressionSql';

/*
  selection array starts on seeing a "SELECT"

  TODO: handle subqueries: https://github.com/uladkasach/sql-code-generator/issues/2
*/
export const extractSelectExpressionsFromQuerySql = ({ sql }: { sql: string }) => {
  // 1. grab the content of everything between SELECT and FROM
  const everythingAfterSelect = (() => {
    const partsSplitOnSelect = sql.split(/(?:SELECT|select)/g);
    if (partsSplitOnSelect.length === 1) throw new Error('no "select" keyword found; unexpected'); // fail fast
    if (partsSplitOnSelect.length > 2) throw new Error('more than one "select" keyword found; not yet supported'); // TODO: https://github.com/uladkasach/sql-code-generator/issues/2
    return partsSplitOnSelect[1];
  })();
  const everythingBetweenSelectAndFrom = (() => {
    const partsSplitOnFromAfterSelect = everythingAfterSelect.split(/(?:FROM|from)/g);
    if (partsSplitOnFromAfterSelect.length === 1) throw new Error('no "from" keyword found; unexpected'); // fail fast
    if (partsSplitOnFromAfterSelect.length > 2) {
      throw new Error('more than one "from" keyword found; not yet supported'); // TODO: https://github.com/uladkasach/sql-code-generator/issues/2
    }
    return partsSplitOnFromAfterSelect[0];
  })();

  // 2. split everything between select and from on commas - they are our select expressions
  const selectExpressionsSql = everythingBetweenSelectAndFrom.split(',').map((thisSql) => thisSql.trim());

  // 3. case each select expression sql into a type def of QuerySelectExpression
  return selectExpressionsSql.map((thisSql) => extractTypeDefinitionFromSelectExpressionSql({ sql: thisSql }));
};
