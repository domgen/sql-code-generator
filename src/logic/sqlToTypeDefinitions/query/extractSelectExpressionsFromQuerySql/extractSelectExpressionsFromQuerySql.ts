import strip from 'sql-strip-comments';

import { castCommasInParensToPipesForTokenSafety } from '../../resource/common/castCommasInParensToPipesForTokenSafety';
import { flattenSqlByReferencingAndTokenizingSubqueries } from '../common/flattenSqlByReferencingAndTokenizingSubqueries';
import { extractTypeDefinitionFromSelectExpressionSql } from './extractTypeDefinitionFromSelectExpressionSql';

/*
  selection array starts on seeing a "SELECT"
*/
export const extractSelectExpressionsFromQuerySql = ({
  sql,
  inASubquery = false, // false by default
}: {
  sql: string;
  inASubquery?: boolean; // used to indicate whether we are currently running on a subquery or not; being in subquery gives us some implicit info
}) => {
  // 0. strip sql comments
  const strippedSql = strip(sql);

  // 1. tokenize each subquery
  const { flattenedSql, references: subqueries } =
    flattenSqlByReferencingAndTokenizingSubqueries({ sql: strippedSql });

  // 2. grab the content of everything between SELECT and FROM
  const everythingAfterSelect = (() => {
    const partsSplitOnSelect = flattenedSql.split(/(?:SELECT|select)/g);
    if (partsSplitOnSelect.length === 1)
      throw new Error('no "select" keyword found; unexpected'); // fail fast
    if (partsSplitOnSelect.length > 2)
      throw new Error('more than one "select" keyword found; unexpected'); // should have been flattened out
    return partsSplitOnSelect[1];
  })();
  const everythingBetweenSelectAndFrom = (() => {
    const partsSplitOnFromAfterSelect = everythingAfterSelect.split(
      /[\n\r\s]+from[\n\r\s]+/gi,
    );
    if (partsSplitOnFromAfterSelect.length === 1) return everythingAfterSelect; // if no "FROM" keyword, then expect that the whole query is just the select expressions
    if (partsSplitOnFromAfterSelect.length > 2) {
      throw new Error('more than one "from" keyword found; unexpected'); // should have been flattened out
    }
    return partsSplitOnFromAfterSelect[0];
  })();

  // 3. cast commas inside of parens into pipes, so that we treat them as unique tokens when splitting "property lines" by comma
  const parenTokenizedInnerSql = castCommasInParensToPipesForTokenSafety({
    sql: everythingBetweenSelectAndFrom,
  });

  // 4. split everything between select and from on commas - they are our select expressions
  const selectExpressionsSql = parenTokenizedInnerSql
    .split(',')
    .map((thisSql) => thisSql.trim());

  // 5. case each select expression sql into a type def of QuerySelectExpression
  return selectExpressionsSql.map((thisSql) =>
    extractTypeDefinitionFromSelectExpressionSql({
      sql: thisSql,
      subqueries,
      inASubquery,
    }),
  );
};
