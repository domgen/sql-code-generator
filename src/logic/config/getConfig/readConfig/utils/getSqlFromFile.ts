import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
