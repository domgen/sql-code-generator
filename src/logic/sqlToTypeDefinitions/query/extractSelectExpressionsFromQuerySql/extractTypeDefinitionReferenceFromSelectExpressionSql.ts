import { SqlSubqueryReference } from '../../../../model/valueObjects/SqlSubqueryReference';
import { TypeDefinitionReference } from '../../../../model/valueObjects/TypeDefinitionReference';
import { throwErrorIfTableReferencePathImpliesTable } from '../common/throwErrorIfTableReferencePathImpliesTable';
import { extractTypeDefinitionReferenceFromSubqueryReferenceToken } from './extractTypeDefinitionReferenceFromSubqueryReferenceToken';

export const extractTypeDefinitionReferenceFromSelectExpression = ({
  sql,
  subqueries,
}: {
  sql: string;
  subqueries: SqlSubqueryReference[];
}) => {
  // 0. try to extract the reference from a subquery, if query references subquery
  const [
    _, // tslint:disable-line no-unused
    subqueryReferenceToken,
  ] = new RegExp(/^(__SSQ:[\w-]+__)(?:\s+(?:[\w\s]*))?$/).exec(sql) ?? [];
  if (subqueryReferenceToken) {
    return extractTypeDefinitionReferenceFromSubqueryReferenceToken({ subqueryReferenceToken, subqueries });
  }

  // 1. try to extract a tableReferencePath, if query references table
  const [
    __, // tslint:disable-line no-unused
    tableReferencePath,
  ] = new RegExp(/^(\w+\.?\w*)(?:\s+(?:[\w\s]*))?$/).exec(sql) ?? [];
  if (tableReferencePath) {
    throwErrorIfTableReferencePathImpliesTable({ referencePath: tableReferencePath }); // check that table reference is explicitly defined, since its best practice
    return new TypeDefinitionReference({
      tableReferencePath,
      functionReferencePath: null,
    });
  }

  // 2. try to extract a functionReferencePath, if query references function
  const [
    ___, // tslint:disable-line no-unused
    referencedFunctionName,
  ] = new RegExp(/(\w+)(?:\([\w\s\.:,\"\|]+\))/).exec(sql) ?? [];
  if (referencedFunctionName) {
    return new TypeDefinitionReference({
      tableReferencePath: null,
      functionReferencePath: `${referencedFunctionName.toLowerCase()}.output`,
    });
  }

  // 3. if couldn't determine type reference with either of the above, throw an error
  throw new Error(`could not extract type definition reference from '${sql}'`);
};
