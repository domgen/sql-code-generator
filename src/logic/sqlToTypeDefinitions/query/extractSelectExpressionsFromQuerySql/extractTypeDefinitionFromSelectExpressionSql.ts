import { TypeDefinitionOfQuerySelectExpression } from '../../../../domain';
import { SqlSubqueryReference } from '../../../../domain/objects/SqlSubqueryReference';
import { extractTypeDefinitionReferenceFromSelectExpression } from './extractTypeDefinitionReferenceFromSelectExpressionSql';

export const extractTypeDefinitionFromSelectExpressionSql = ({
  sql,
  subqueries,
  inASubquery,
}: {
  sql: string;
  subqueries: SqlSubqueryReference[];
  inASubquery: boolean; // used to indicate whether we are currently running on a subquery or not; being in subquery gives us some implicit info
}) => {
  // grab the type reference
  const typeReference = extractTypeDefinitionReferenceFromSelectExpression({
    sql,
    subqueries,
  });

  // grab the alias, if any
  const [__, specifiedAlias] =
    new RegExp(/(?:[\w\.,:\(\)\s\|\']+)(?:\s+)(?:as|AS)(?:\s+)(\w+)/g).exec(
      sql,
    ) ?? []; // tslint:disable-line no-unused

  // define the alias, considering whether alias was specified
  if (
    !specifiedAlias &&
    !!typeReference.functionReferencePath &&
    !inASubquery
  ) {
    // throw error if fn reference and alias not defined, as its best practice to explicitly define it - and we're not going to "infer" it for the user
    throw new Error(
      `
select expressions that reference a function must have an alias defined, per best practice.
  - e.g., \`select concat(n.first, n.last)\` => \`select concat(n.first, n.last) as full_name\`

\`${typeReference.functionReferencePath}\` does not meet this criteria.
`.trim(),
    );
  }
  const alias = (() => {
    if (specifiedAlias) return specifiedAlias;
    if (inASubquery) return '__subquery_placeholder_name__'; // we wont use this alias anywhere, since we're in a subquery and subquery just returns one value; therefore, we can give it a fake name as a placeholder
    if (typeReference.tableReferencePath)
      return typeReference.tableReferencePath.split('.').slice(-1)[0]!;
    throw new Error('could not define alias for sql expression; unexpected');
  })();

  // return the full definition
  return new TypeDefinitionOfQuerySelectExpression({
    alias,
    typeReference,
  });
};
