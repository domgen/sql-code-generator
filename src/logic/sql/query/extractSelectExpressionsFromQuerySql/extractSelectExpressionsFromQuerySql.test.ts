import { extractSelectExpressionsFromQuerySql } from './extractSelectExpressionsFromQuerySql';
import { getSqlFromFile } from '../../../config/_utils/getSqlFromFile';

describe('extractSelectExpressionsFromQuerySql', () => {
  it('should be able to determine types accurately for this example', async () => {
    const sql = await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/selectImageById.sql` });
    const defs = extractSelectExpressionsFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example', async () => {
    const sql = await getSqlFromFile({ filePath: `${__dirname}/../__test_assets__/selectSuggestion.sql` });
    const defs = extractSelectExpressionsFromQuerySql({ sql });
    expect(defs.slice(-1)[0].name).toEqual('updated_at');
    expect(defs.slice(-1)[0].sourcePath).toEqual('v.created_at');
    expect(defs).toMatchSnapshot();
  });
});
