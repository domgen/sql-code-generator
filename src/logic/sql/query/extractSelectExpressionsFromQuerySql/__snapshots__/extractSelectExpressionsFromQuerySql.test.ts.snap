// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this example 1`] = `
Array [
  TypeDefinitionOfQuerySelectExpression {
    "name": "id",
    "sourcePath": "id",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "url",
    "sourcePath": "url",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "caption",
    "sourcePath": "caption",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "alt_text",
    "sourcePath": "alt_text",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "credit",
    "sourcePath": "credit",
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this other example 1`] = `
Array [
  TypeDefinitionOfQuerySelectExpression {
    "name": "id",
    "sourcePath": "s.id",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "uuid",
    "sourcePath": "s.uuid",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "suggestion_source",
    "sourcePath": "s.suggestion_source",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "external_id",
    "sourcePath": "s.external_id",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "suggested_idea_id",
    "sourcePath": "s.suggested_idea_id",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "resultant_curated_idea_id",
    "sourcePath": "s.resultant_curated_idea_id",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "status",
    "sourcePath": "v.status",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "result",
    "sourcePath": "v.result",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "created_at",
    "sourcePath": "s.created_at",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "effective_at",
    "sourcePath": "v.effective_at",
  },
  TypeDefinitionOfQuerySelectExpression {
    "name": "updated_at",
    "sourcePath": "v.created_at",
  },
]
`;
